File: dotnet/Gherkin/ParseError.cs
namespace Gherkin
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    public class ParseException : Exception 
    {
        public string State { get; private set; }
        public IEnumerable<string> ExpectedEvents { get; private set; }

        public ParseException(string state, string @event, IEnumerable<string> expectedEvents, Position position)
            : base("Parse error on line " + position.Line + ". Found " + @event + " when expecting one of: " + string.Join(", ", expectedEvents.ToArray()) + ". (Current state: " + state + ").")
        {
            State = state;
            ExpectedEvents = expectedEvents;
        }
    }
}

File: java/src/main/java/gherkin/parser/ParseError.java
package gherkin.parser;

import gherkin.util.FixJava;

import java.util.List;

public class ParseError extends RuntimeException {
    private final String state;
    private final List<String> legalEvents;

    public ParseError(String state, String event, List<String> legalEvents, String uri, int line) {
        super("Parse error at " + uri + ":" + line + ". Found " + event + " when expecting one of: " + FixJava.join(legalEvents, ", ") + ". (Current getState: " + state + ").");
        this.state = state;
        this.legalEvents = legalEvents;
    }

    public List<String> getLegalEvents() {
        return legalEvents;
    }

    public String getState() {
        return state;
    }
}
